OpenImageIOReader : Don't Use the OpenImageIO Cache

danieldresser-ie opened the pull request:
This should be a nice step forward for loading images in Gaffer.  By loading images directly into the Gaffer cache, we avoid repeating the conversion from OIIO data layout, and the user only has to worry about the Gaffer cache size, instead of tuning two separate caches.  It appears to improve performance in some reasonable read-bound compositing Gaffer scripts by about 20%.

This should be pretty good to look at least - I haven't yet tested quite as much as I was planning to, because some of the internal projects I'd like to test haven't been updated to work with Gaffer master yet ( hopefully this will happen tomorrow after the imath binding PR gets merged ).

The Offest and Merge fixes aren't technically related to the core of this PR, but those bugs were uncovered by adding proper error handling to the reader.

The contentious part is the performance issue we talked about earlier, and the last two commits.  Using a mutex on each file handle to serialize all access to the tileBatch plug for a particular file does make sure two threads don't repeat loading the data, but it also slows down some accesses that could proceed multithreaded, because the tileBatch is already cached.

I have tried to do some testing of how bad this could get.  In cases where there is no heavy contention on the cache, it's certainly no problem.  In cases where the cache is heavily contended, it appears to be possible for it to get almost arbitrarily bad.  It's pretty hard to put together a test case which is simple, realistic, and heavily contended ( in production, the cache would be large, and if this came up, it would be in complex scenes ).  But I've put together a very simple test, which I will paste below.  If you set the Gaffer cache size to 100 megs, and run this:

`gaffer stats ~/test_crud/readPerfContended.gfr -image ImageLoop.out`

It takes 4 seconds in this PR, but without the last two commits, it takes 40 - 50 seconds.

My current assumption about the mechanism for this to get so bad is that thread A could need a tileBatch that is currently in the cache, but instead of just taking it, it locks on the file mutex.  Thread B that has the file mutex is loading a different piece of the file, and by loading it, it fills the cache and  evicts the tileBatch that thread A needed.  So when thread A finally gets in, it has to reload from disk, potentially triggering further lengthy locks and evictions.  The number of actual evaluations of ImageReader1.__oiioReader.__tileBatch in the performance stats goes from 900 to 9000 in the performance stats.

Maybe this is OK to merge without the last two commits if we just assume that we must always make the cache large enough that heavy contention doesn't occur, but that feels like a dangerous assumption to make.   I feel like there has to be some way of making the extra little piece of API needed small and explicit enough that it isn't that big a loss of cleanliness ( and we should be able to come back and delete it once we come up with something better ).

But anyway, at least you can look through the rest of it and make sure it looks OK.  Hopefully tomorrow I'll actually manage to read some deep images now that I've got this in place.

danieldresser-ie commented:
Paste of readPerfContended.gfr - note half of this is just to generate the test file.  Run ImageWriter once to create /tmp/3k.exr, and the the test should be usable in gaffer stats.

```
import Gaffer
import GafferDispatch
import GafferImage
import IECore

__children = {}

__children["ImageLoop"] = GafferImage.ImageLoop( "ImageLoop" )
parent.addChild( __children["ImageLoop"] )
__children["ImageLoop"]["iterations"].setValue( 5 )
__children["Merge"] = GafferImage.Merge( "Merge" )
parent.addChild( __children["Merge"] )
__children["Merge"]["in"].addChild( GafferImage.ImagePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Offset"] = GafferImage.Offset( "Offset" )
parent.addChild( __children["Offset"] )
__children["Expression1"] = Gaffer.Expression( "Expression1" )
parent.addChild( __children["Expression1"] )
__children["Expression1"]["__out"].addChild( Gaffer.V2iPlug( "p0", direction = Gaffer.Plug.Direction.Out, defaultValue = IECore.V2i( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Constant"] = GafferImage.Constant( "Constant" )
parent.addChild( __children["Constant"] )
__children["Constant"]["format"].setValue( GafferImage.Format( 1, 1, 1.000 ) )
__children["Constant"]["color"].setValue( IECore.Color4f( 1, 0.5, 0.25, 1 ) )
__children["Crop"] = GafferImage.Crop( "Crop" )
parent.addChild( __children["Crop"] )
__children["Crop"]["area"].setValue( IECore.Box2i( IECore.V2i( 0, 0 ), IECore.V2i( 2, 2 ) ) )
__children["Blur"] = GafferImage.Blur( "Blur" )
parent.addChild( __children["Blur"] )
__children["Blur"]["radius"]["x"].setValue( 1.0 )
__children["Blur"]["expandDataWindow"].setValue( True )
__children["Resize"] = GafferImage.Resize( "Resize" )
parent.addChild( __children["Resize"] )
__children["Resize"]["format"].setValue( GafferImage.Format( 3400, 2200, 1.000 ) )
__children["Resize"]["fitMode"].setValue( 4 )
__children["Resize"]["filter"].setValue( 'triangle' )
__children["ImageWriter"] = GafferImage.ImageWriter( "ImageWriter" )
parent.addChild( __children["ImageWriter"] )
__children["ImageWriter"]["fileName"].setValue( '/tmp/3k.exr' )
__children["ImageReader1"] = GafferImage.ImageReader( "ImageReader1" )
parent.addChild( __children["ImageReader1"] )
__children["ImageReader1"]["fileName"].setValue( '/tmp/3k.exr' )
__children["ImageReader1"]["refreshCount"].setValue( 4 )
__children["ImageLoop"]["in"].setInput( __children["Offset"]["out"] )
__children["ImageLoop"]["next"].setInput( __children["Merge"]["out"] )
__children["Merge"]["in"]["in0"].setInput( __children["ImageLoop"]["previous"] )
__children["Merge"]["in"]["in1"].setInput( __children["Offset"]["out"] )
__children["Offset"]["in"].setInput( __children["ImageReader1"]["out"] )
__children["Offset"]["offset"].setInput( __children["Expression1"]["__out"]["p0"] )
__children["Crop"]["in"].setInput( __children["Blur"]["out"] )
__children["Blur"]["in"].setInput( __children["Constant"]["out"] )
__children["Blur"]["radius"]["y"].setInput( __children["Blur"]["radius"]["x"] )
__children["Resize"]["in"].setInput( __children["Crop"]["out"] )
__children["ImageWriter"]["in"].setInput( __children["Resize"]["out"] )
__children["Expression1"]["__engine"].setValue( 'python' )
__children["Expression1"]["__expression"].setValue( 'parent["__out"]["p0"] = IECore.V2i( context.get( "loop:index", 1 ) * 300 )' )


del __children
```

johnhaddon commented:
I wanted to start this review by taking a look at the performance/caching issues, and it turned out to be quite interesting on a couple of counts. While it's fresh in my head I wanted to write it down - I'll do a second review which actually looks at the code later.

Performance
-----------------

I'm seeing a performance improvement here too - about a 30% reduction for your readPerfContended.gfr example (with the default cache size). Good stuff!

OIIO Caching
------------------

I wanted to see if I could reproduce a performance drop-off using the code from master, but with an unrealistically small OIIO cache. My naive hope was that a 100Mb OIIO cache on master would give similar results to a 100Mb Gaffer from this PR, and then we could ignore the questions posed by the last couple of commits.

I couldn't reproduce any performance drop off _at all_ at first. It seems that OIIO's cache treats scanline images as having a single huge tile, and is willing to blow through its cache limits to store it. I recall some discussion of just how horrific this was going to be for deep images if they were ever supported by OIIO's cache, so that's definitely another validation of the new approach we're taking.

Switching to a tiled image allowed me to at least get OIIO to evict something. But even then, limiting the cache to 100Mb basically had no effect on runtime. This makes sense though, since the image itself is only a smidge over the limit (114Mb), and the OIIO cache isn't mixed with Gaffer's, so no downstream computations can evict anything from the cache. With this PR, the same cache is used for both loading the images and for processing them, so downstream stuff can evict upstream stuff.

I then tried an OIIO cache of 10Mb (the smallest OIIO will allow), and was at last able to reproduce some performance drop off - about a 50% increase in runtime. So, nothing anywhere near as dramatic as your worst case scenario with the new caching. But then again, perhaps this is because the cache isn't shared? Do we think there might be any benefit in isolating parts of Gaffer's cache in different memory pools? So image loading is separate from processing? Or scene stuff is separate from image stuff? I'm not sure, but it might be something to ponder in the future...

Cache thrashing
---------------------

Next I tried to reproduce your worst case example from this PR, where you saw an increase from 4 seconds to 40 seconds. I switched back to using a scanline file for this, so my results were more comparable to yours. These are the timings I got :

|   | Default Cache | 100Mb Cache | 100Mb Cache, dropping last two commits
|--|--------------------|--------------------|--------------------------------------------------------
Time | 1.276s | 8.152s | 71.139s
Relative Time | 1 | 6.3 | 55

So, even worse than what you saw! I did also try a tiled image for kicks, and it was similarly bad.

ImageAlgo::parallelProcessTiles, our villain
--------------------------------------------------------

Seeing how severe the thrashing was got me to thinking about tile access patterns, so I took a little look at what `gaffer stats -image` is doing. It uses `ImageAlgo::parallelProcessTiles` to visit every tile/channel pair, pulling on the channelData plug to compute the tile. Internally, `parallelProcessTiles` is using `tbb::parallel_for` with a 3d range (x, y, channel). It turns out that this is _awful_ - there's no guarantee of any particular execution order, and although TBB claims to process things in somewhat coherent order, in practice there is none of the scanline-by-scanline coherency that we would benefit from - things might be coherent in the other dimensions instead.

So I switched to using `ImageAlgo::parallelGatherTiles` because it uses a `tbb::pipeline` internally, allowing it to process everything in roughly scanline order. With this, the 71s worst case dropped to only 2s! This seems much more in line with what we might hope for even when we know our cache is too small.

We're using `parallelProcessTiles` in the viewer, so we need to get this fixed ASAP - I think probably the easiest thing is to just reimplement it to call `parallelGatherTiles` internally with a null gather step.

Those last two commits
-------------------------------

As you know, I don't really want to expose the internals of the cache with the new `ValuePlug::getFromCache()` method. Instead I want to push forward with improving the cache internals so there's no need for it, and so the internals can remain internal. I won't be too dogmatic about that if there is a huge performance win to be had, but I'm hoping that my "gather vs process" results above might convince you that it's not necessary just now - provided we stick with reasonably coherent access patterns, the cache behaves reasonably even when given very little memory to play with.

I propose that we push ahead without the last two commits, but if we hit genuine issues later with production-worthy cache sizes (and I still haven't got my cache improvements rolled out), we reconsider. While I'm thinking about it, here's how I think we could make the `getFromCache()` approach a bit more palatable :

- Don't expose the internal IECore::Object via a public method on ValuePlug. This is deliberately hidden from public view because :
     - Exposing it would be confusing for simple types like NumericPlug. We made this mistake for `IECore::Parameter` and `IECore::NumericParameter` and have had a confusing API ever since.
     - It provides us greater latitude to change implementations later - for instance, I have some potential StringPlug optimisations which would necessitate storing something other than StringData internally.
   - For compound ValuePlugs (V3fPlug for instance), there is no Object anyway - they're only stored at leaf plugs. We should shield people from this sort of implementation detail.
- Instead only expose the functionality via TypedObjectPlug, where we know we're a leaf plug, and the caller is expecting to deal with Objects anyway.
- Perhaps just implement it via an optional `nullIfUncached` argument to `TypedObjectPlug::getValue()`?

johnhaddon commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -323,6 +562,94 @@ void OpenImageIOReader::compute( ValuePlug *output, const Context *context ) con
+		// at this point
+		Box2i dataRegion = file->readRegion( tileBatchIndex.z, targetRegion, fileData );
+		
+		const int nchannels = file->m_imageSpec.nchannels;
```
Again, is this really valid for all subimages?

johnhaddon commented on python/GafferImageTest/OpenImageIOReaderTest.py:
```diff
@@ -169,9 +169,8 @@ def testNoOIIOErrorBufferOverflows( self ) :
-		for i in range( 0, 300000 ) :
-			with IECore.IgnoredExceptions( Exception ) :
-				n["out"]["dataWindow"].getValue()
+		with IECore.IgnoredExceptions( Exception ) :
```
If you're sure it's safe to remove this test, then let's just remove it completely, because without the loop it's useless. It was guarding against a scenario where we never retrieved the error message from the OIIO cache, and multiple errors just accumulated and then went over an internal limit...

johnhaddon commented on src/GafferImage/Offset.cpp:
```diff
@@ -167,8 +167,12 @@ IECore::ConstFloatVectorDataPtr Offset::computeChannelData( const std::string &c
 	}
 	else
 	{
+		Box2i inDataWindow = inPlug()->dataWindowPlug()->getValue();
```
I feel like I'm missing something here. If the output data window is derived from the input data window, and downstream nodes must only pull tiles from the output data window, why do we need to do any guarding? As long as the downstream nodes are keeping their promise, doesn't that mean we are too?

If there is a genuine bug here though, then `hashChannelData()` would need updating to reflect the changes in `computeChannelData()`.

danieldresser-ie replied:
We previously took the bound of the entire output tile, and shifted it into input space.  The output could have as little as one pixel overlapping with the actual output data window.  The tiles are aligned differently in input space, so of the 4 input tiles necessary to compute the output tile, it's possible that only one is valid.  I will add the update to hashChannelData.

johnhaddon replied:
Thanks for the explanation.

johnhaddon commented on src/GafferImage/Merge.cpp:
```diff
@@ -308,7 +308,9 @@ IECore::ConstFloatVectorDataPtr Merge::merge( F f, const std::string &channelNam
-		if( ImageAlgo::channelExists( channelNames, channelName ) )
+		const Box2i validBound = boxIntersection( tileBound, dataWindow );
+
+		if( ImageAlgo::channelExists( channelNames, channelName ) && !BufferAlgo::empty( validBound ) )
```
We need to update the matching tests in `hashChannelData()` so that they continue to match. Since we seem to be relying on the OpenImageIOReader to detect these problems, do you think it's worth updating `OpenImageIOReader::hashChannelData()` to throw the same error that `OpenImageIO::computeChannelData()` is throwing?

johnhaddon commented on python/GafferImageTest/OpenImageIOReaderTest.py:
```diff
@@ -488,5 +490,42 @@ def testFileFormatMetadata( self ) :
+				w.execute()
+				rBack['refreshCount'].setValue( rBack['refreshCount'].getValue() + 1 )
+
+				self.assertImagesEqual( r["out"], offsetIn["out"], ignoreMetadata = True ) #maxDifference = 0.001 )
```
Remove dead code (`#maxDifference`).

johnhaddon commented on python/GafferImageTest/ImageWriterTest.py:
```diff
@@ -368,7 +368,7 @@ def testMultipleWrite( self ) :
 		r["fileName"].setValue( self.__largeFilePath )
 
-		for i in range( 20 ) :
+		for i in range( 5 ) :
```
We can just remove the entire test instead - it was related to the OIIO ImageCache that we're no longer using. When removing it, could you include a reference to https://github.com/GafferHQ/gaffer/issues/1923, since that has a bunch of relevant history...

danieldresser-ie commented:
> Since we seem to be relying on the OpenImageIOReader to detect these problems, do you think it's worth updating OpenImageIOReader::hashChannelData() to throw the same error that OpenImageIO::computeChannelData() is throwing?

Hmm, currently the hash only depends on hashing inputs, and doesn't have to do any getValue()s.  I don't really like the idea of adding a getValue() on the dataWindow just to catch internal bugs.  I am doing some testing of catching invalid accesses in the unit tests with this check in place, but I'm not sure I want to leave it in the code long term.

Between formatting fixes, and hunting down newly revealed unit tests failures, I didn't get through as much as I was hoping today.

Those are some good catches on the subimage stuff.  I feel like it is a lot easier to think about now while this stuff is fresh, so I think I will add some tests and try cleaning it up.

As for the performance things: parallelProcessTiles sounds like a very good thing to get fixed.  Do you want me to look at that?

I still feel like it's a bit of a separate issue from trying to insure we don't perform horribly under hash contention though?  It would just mean that our hash contention test would have to add a rotation in the chain in order to trigger a bad access pattern?  I probably have gotten frustrated enough with trying to get this ready to go and move onto the deep stuff that I will give up and leave the performance hazard alone - though I do still feel like the performance hazard is a much uglier thing to leave lying around in the codebase then a clearly documented special case accessor function that isn't intended for public use.

johnhaddon commented:
> Hmm, currently the hash only depends on hashing inputs, and doesn't have to do any getValue()s. I don't really like the idea of adding a getValue() on the dataWindow just to catch internal bugs. I am doing some testing of catching invalid accesses in the unit tests with this check in place, but I'm not sure I want to leave it in the code long term.

OK.

> Those are some good catches on the subimage stuff. I feel like it is a lot easier to think about now while this stuff is fresh, so I think I will add some tests and try cleaning it up.

OK - let's give that a try then. My only worry is the comments I made about `seek_subimage()` - if seek performance _is_ an issue, then I wonder if we'd need to structure things differently, maybe even using an ImageInput per subimage or something like that. Would have been nice to deal with this separately from the caching...

> As for the performance things: parallelProcessTiles sounds like a very good thing to get fixed. Do you want me to look at that?

I can take a look if you prefer - just let me know...

> I still feel like it's a bit of a separate issue from trying to insure we don't perform horribly under hash contention though? It would just mean that our hash contention test would have to add a rotation in the chain in order to trigger a bad access pattern?

Yeah, true.

> I probably have gotten frustrated enough with trying to get this ready to go and move onto the deep stuff that I will give up and leave the performance hazard alone - though I do still feel like the performance hazard is a much uglier thing to leave lying around in the codebase then a clearly documented special case accessor function that isn't intended for public use.

My concern is always that things will get used regardless of documentation etc, and then become a block to refactoring. Maybe I'm emphasising that too much in this instance though. If you could rejig it along the lines I suggested in my first review, then that'd be great...

danieldresser-ie commented:
OK, I think that's all the inline comments addressed.  There is a bunch more documentation around OpenImageIOReader::File now, so take a look at that and see if it's clearer.

I added some basic tests for multipart images.  I'm not too worried about performance of multipart until it becomes an issue - I did do a quick trace of what code seek_subimage is actually calling on an EXR:
https://github.com/OpenImageIO/oiio/blob/master/src/openexr.imageio/exrinput.cpp#L826

My conclusions were:
* there's a very early out if the subimage hasn't changed ( so it's not hurting the non-multipart case )
* it does some arithmetic to set up the spec, but the only potentially expensive thing appears to be instantiating Imf::*InputPart, which just calls this:
https://github.com/ImageMagick/exr/blob/master/openexr/IlmImf/ImfMultiPartInputFile.cpp#L180
... which is just taking a lock and doing a map lookup if it's already cached.
Tile batches are large enough that we're not doing this a massive number of times ( about 32 * <number of parts> for a 4K image ), so it doesn't seem that concerning.  I'm not expecting the multipart stuff to get hammered too hard right away, my only reason for wanting to keep it was that I had to think about it while dealing with the OIIO API, and it seemed nicer to leave it with the basics working in case we do have to come along and implement stereo support or something ( which would definitely require further work ).

That just leaves those last two commits.  I had a go at implementing what it sounded like you wanted.  This involved making TypedObjectPlug a friend of ValuePlug, and too me feels a bit uglier ( If we have a better plan coming soon, the raw getFromCache call is very easy to deprecate - it can just always return null, and seems very easy to rip out.  If you're concerned about people ignoring the docs and using it, we could have just given it a scarier name. ) But anyway, if this is what you want it to look like, seems OK to me.  I would just need to squash it and update the commenting around it.

danieldresser-ie commented:
OK, I've fixed up that error message ( OIIO does automatically include the filename in both cases, so I've removed the filename from our part of the error message ), and I've included your getValueIfCached commit ( my only changes were squashing your two commits, and removing one comment from ValuePlug.h which referenced the cachedOnly parameter which no longer exists on getObjectValue ).

After switching OpenImageIOReader to use that, I'm seeing the same performance gain in  readPerfContended.gfr, so looks like this is good to go.

danieldresser-ie commented:
To address an aside from earlier before we get this merged and forget this conversation:

I started taking a look at your proposal to reimplement parallelProcessTiles as a call to parallelGatherTiles with a null gather step ... but then realized all this tbb stuff looks really complicated - if you're familiar enough with it that this seems straightforward to you, I should probably let you do it.

I should mention that I reproduced your findings on the performance benefit of your commit that switched processTiles to use parallelGatherTiles in the heavily contended case.  I also noticed a reproducible 5% decrease in performance in a non-contended case though - not sure what's up with that.

danieldresser-ie commented:
While starting to look at loading deep images, I realized one last thing I should fix here before you merge it - the File class was originally a simple struct, and it hasn't been converted to a class since becoming more complex.  I should probably make some stuff private on it, and add a const accessor on m_imageSpec.  Should be quick, but I should probably wait until Monday when I'm more alert.

danieldresser-ie commented:
OK, did that final bit of clean-up mentioned in the last comment.  Hopefully this is good to go now.

johnhaddon commented:
Great - thanks Daniel! I've merged, but note that this was _after_ tagging and releasing 0.42. I don't anticipate any issues with this PR at all, but while compiling the release notes I became wary of just how much was being crammed in to this one version, and opted for caution. I'm hoping we'll be able to get the deep support ready for the next release, at which point this work will really come into its own.
Cheers...
John
